### PR TITLE
Connection string logic and authentication

### PR DESCRIPTION
**Pull request by katiecai**
Adding logic for configuring output/trigger with a connection string instead of individual host/user/pw/port parameters. Needs some cleanup

For the output implementation, I decided to include all params PLUS the connection string as optional attributes and then let the RabbitMQ API handle any missing fields. I think this makes the most sense as I would have to do a lot of unnecessary error checking otherwise.

How to design the trigger is a little trickier. The RabbitMQ API has defaults for any missing fields so the user should essentially be able to pass in whatever parameters they wish to, but this makes the trigger code a lot more complicated if we want to account for all of these combinations. Because I have to create trigger objects, I thought about writing a bunch of overloaded constructors to handle different combinations of filled in/missing fields. I've written a few overloaded constructors in this PR but I wanted some feedback on whether or not this is the best way to handle this issue because then I'd have to create like 10+ constructors. 

Also: closing https://github.com/Azure/azure-functions-rabbitmq-extension/pull/27 to combine authentication and connection string logic into one PR

Edit: Adding requeuing logic here as well since it's not a lot of code. Will need to verify whether this is exactly what we want to do but according to [documentation](https://www.rabbitmq.com/confirms.html#consumer-nacks-requeue) this seems to be the solution we want?

Edit 2: The NACK logic seems to requeue over and over even when something hasn't gone wrong. Some say the default behavior is that the message won't leave the queue if there's an error. I need to confirm this but if it's true we don't need to add any other logic for resending

### CODE REVIEW

File: src/RabbitMQAttribute.cs [hunk 1]
```diff
@@ -29,6 +29,17 @@ public sealed class RabbitMQAttribute : Attribute
...
+        [AutoResolve]
+        public string Password { get; set; }
+
+        // Optional
+        public int Port { get; set; }
+
+        public string ConnectionStringSetting { get; set; }
+
         [AutoResolve]
         public string Exchange { get; set; }
```
Comment (pragnagopa) [inline src/RabbitMQAttribute.cs]: Is exchange optional? If yes, add documentation in summary tags with info on the default value

File: src/Trigger/RabbitMQListener.cs [hunk 2]
```diff
@@ -59,6 +60,7 @@ public Task StartAsync(CancellationToken cancellationToken)
 
             _consumer.Received += (model, ea) =>
             {
+                _channel.BasicAck(ea.DeliveryTag, true);
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQListener.cs]: is this an intentional change for this PR? Will you be addressing Acks on a separate PR?

File: src/Trigger/RabbitMQTriggerAttribute.cs [hunk 3]
```diff
@@ -9,24 +9,39 @@ namespace Microsoft.Azure.WebJobs
...
-        public RabbitMQTriggerAttribute(string hostName, string queueName, ushort batchNumber)
+        public RabbitMQTriggerAttribute(string hostName, string userName, string password, int port, string queueName)
         {
             this.HostName = hostName;
+            this.UserName = userName;
+            this.Password = password;
+            this.Port = port;
             this.QueueName = queueName;
-            this.BatchNumber = batchNumber;
+            this.BatchNumber = 1;
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttribute.cs]: Can BatchNumber be set by users?

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs [hunk 4]
```diff
@@ -39,13 +39,33 @@ public Task<ITriggerBinding> TryCreateAsync(TriggerBindingProviderContext contex
...
+            string userName = Resolve(attribute.UserName);
+
+            string password = Resolve(attribute.Password);
+
+            int port = attribute.Port;
+
             ushort batchNumber = attribute.BatchNumber;
 
-            IRabbitMQService service = _provider.GetService(hostName, queueName);
+            if (string.IsNullOrEmpty(connectionString) && !string.IsNullOrEmpty(hostName) && ((string.IsNullOrEmpty(userName) && hostName != "localhost") ||
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: defaulting to localhost if hostName is not specified will make this check simpler. Also, move this the condition check to Utility and add Unit tests

File: test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQSamples.cs [hunk 5]
```diff
@@ -16,7 +16,7 @@ public static class RabbitMQSamples
         public static void TimerTrigger_StringOutput(
             [TimerTrigger("00:01")] TimerInfo timer,
             [RabbitMQ(
-                HostName = "localhost",
+                ConnectionStringSetting = "amqp://guest:guest@localhost:5672/",
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQSamples.cs]: do not give connection string here. this needs to be an environment variable

File: test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQSamples.cs [hunk 6]
```diff
@@ -76,7 +70,7 @@ public static class RabbitMQSamples
 
         // Trigger samples
         public static void RabbitMQTrigger_String(
-             [RabbitMQTrigger("localhost", "queue")] string message,
+             [RabbitMQTrigger("amqp://guest:guest@localhost:5672/", "queue")] string message,
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQSamples.cs]: Add test that uses username and password. Note: UserName and password should be setting names as well similar to Twilio SID and password

File: test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQSamples.cs [hunk 7]
```diff
@@ -85,21 +79,21 @@ public static class RabbitMQSamples
...
-            [RabbitMQTrigger("localhost", "queue")] TestClass pocObj,
+            [RabbitMQTrigger("queue")] TestClass pocObj,
             ILogger logger)
         {
             logger.LogInformation($"RabbitMQ queue trigger function processed message: {pocObj}");
         }
 
         public static void RabbitMQTrigger_RabbitMQOutput(
-            [RabbitMQTrigger("localhost", "queue")] string inputMessage,
+            [RabbitMQTrigger("queue")] string inputMessage,
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQSamples.cs]: Add comments to indicate defaults to localhost if HostName is not specified.

File: src/RabbitMQAttribute.cs [hunk 8]
```diff
@@ -29,6 +29,17 @@ public sealed class RabbitMQAttribute : Attribute
...
+        [AutoResolve]
+        public string UserName { get; set; }
+
+        [AutoResolve]
+        public string Password { get; set; }
+
+        // Optional
+        public int Port { get; set; }
+
+        public string ConnectionStringSetting { get; set; }
```
[the remark this answers is not in the document]
Comment (pragnagopa) [reply src/RabbitMQAttribute.cs]: Also, Missing attribute AppSetting on Password and Username

File: src/RabbitMQAttribute.cs [hunk 9]
```diff
@@ -29,6 +29,17 @@ public sealed class RabbitMQAttribute : Attribute
         [AutoResolve]
         public string QueueName { get; set; }
```
Comment (pragnagopa) [inline src/RabbitMQAttribute.cs]: Nit: remove this. references across all the files

File: src/Trigger/RabbitMQTriggerAttribute.cs [hunk 10]
```diff
@@ -9,24 +9,39 @@ namespace Microsoft.Azure.WebJobs
     [Binding]
     public sealed class RabbitMQTriggerAttribute : Attribute
     {
-        public RabbitMQTriggerAttribute(string hostName, string queueName)
+        public RabbitMQTriggerAttribute(string connectionString, string queueName)
+        {
+            this.ConnectionStringSetting = connectionString;
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttribute.cs]: remove this.

File: src/Trigger/RabbitMQTriggerAttribute.cs [hunk 11]
```diff
@@ -9,24 +9,39 @@ namespace Microsoft.Azure.WebJobs
...
             this.HostName = hostName;
+            this.UserName = userName;
+            this.Password = password;
+            this.Port = port;
             this.QueueName = queueName;
-            this.BatchNumber = batchNumber;
+            this.BatchNumber = 1;
         }
 
+        public string ConnectionStringSetting { get;  }
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttribute.cs]: Missing ConnectionString attribute

File: src/Trigger/RabbitMQTriggerAttribute.cs [hunk 12]
```diff
@@ -9,24 +9,39 @@ namespace Microsoft.Azure.WebJobs
...
+
         public string HostName { get; }
 
         public string QueueName { get; }
 
+        public string UserName { get; }
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttribute.cs]: Missing AppSetting attribute on username and password

File: src/Config/RabbitMQExtensionConfigProvider.cs [hunk 13]
```diff
@@ -61,36 +61,54 @@ public void Initialize(ExtensionConfigContext context)
 
         public void ValidateBinding(RabbitMQAttribute attribute, Type type)
         {
+            string connectionString = Utility.FirstOrDefault(attribute.ConnectionStringSetting, _options.Value.ConnectionString);
             string hostName = Utility.FirstOrDefault(attribute.HostName, _options.Value.HostName);
```
[the remark this answers is not in the document]
Comment (pragnagopa) [reply src/Config/RabbitMQExtensionConfigProvider.cs]: Add new file constants.cs and move localhost to a constant

File: src/Config/RabbitMQExtensionConfigProvider.cs [hunk 14]
```diff
@@ -61,36 +61,54 @@ public void Initialize(ExtensionConfigContext context)
...
-            string queueName = Utility.FirstOrDefault(attribute.QueueName, _options.Value.QueueName);
-
-            if (string.IsNullOrEmpty(hostName))
-            {
-                throw new InvalidOperationException("RabbitMQ host name is missing");
-            }
+            string queueName = Utility.FirstOrDefault(attribute.QueueName, _options.Value.QueueName) ?? throw new InvalidOperationException("RabbitMQ queue name is missing");
 
-            if (string.IsNullOrEmpty(queueName))
+            if (!string.IsNullOrEmpty(hostName) && string.IsNullOrEmpty(connectionString) && !hostName.Equals("localhost", StringComparison.InvariantCultureIgnoreCase))
```
Comment (pragnagopa) [inline src/Config/RabbitMQExtensionConfigProvider.cs]: Same here: Move the condition to check for localhost to utility and you can reuse that function in TriggerBindingProvider as well

File: src/Services/RabbitMQService.cs [hunk 15]
```diff
@@ -10,15 +10,50 @@ internal sealed class RabbitMQService : IRabbitMQService
...
         {
-            _hostName = hostName ?? throw new ArgumentNullException(nameof(hostName));
+            _connectionString = connectionString;
+            _hostName = hostName;
             _queueName = queueName ?? throw new ArgumentNullException(nameof(queueName));
+            _userName = userName;
+            _password = password;
+            _port = port;
+
+            ConnectionFactory connectionFactory = new ConnectionFactory();
```
Comment (pragnagopa) [inline src/Services/RabbitMQService.cs]: Move logic that builds connectionFactory to a separate method. Add unit tests that verify various if conditions

File: src/Services/RabbitMQService.cs [hunk 16]
```diff
@@ -10,15 +10,50 @@ internal sealed class RabbitMQService : IRabbitMQService
...
+                connectionFactory.Password = _password;
+            }
+
+            if (_port != 0)
+            {
+                connectionFactory.Port = _port;
+            }
 
-            ConnectionFactory connectionFactory = new ConnectionFactory() { HostName = _hostName };
             _channel = connectionFactory.CreateConnection().CreateModel();
```
Comment (pragnagopa) [inline src/Services/RabbitMQService.cs]: Nit: Instead of exposing GetBatch, GetChannel methods, expose getters
```
 public IBasicPublishBatch Batch => _batch;

public IModel Channel => _channel;
```
Change IRabbitMQService to
```
   IModel Channel { get; }

   IBasicPublishBatch Batch { get; }
```

Comment (pragnagopa) [reply src/Services/RabbitMQService.cs]: Also, rename channel to  Model as the description here: https://www.rabbitmq.com/releases/rabbitmq-dotnet-client/v3.4.2/rabbitmq-dotnet-client-3.4.2-client-htmldoc/html/type-RabbitMQ.Client.IModel.html states IModel is interface for AMQP Model

Comment (pragnagopa) [reply src/Services/RabbitMQService.cs]: Rename Batch to BasicPublishBatch. Although BasicPublishBatch is not descriptive, name Batch is misleading

File: src/Trigger/RabbitMQListener.cs (hunk 2 above)
Comment (katiecai) [reply src/Trigger/RabbitMQListener.cs]: I just included it as a part of this PR because the samples don't work as expected otherwise. But more in depth changes would be made in a separate PR

File: src/Trigger/RabbitMQTriggerAttribute.cs (hunk 3 above)
Comment (katiecai) [reply src/Trigger/RabbitMQTriggerAttribute.cs]: I used to have an overloaded constructor that takes in batch number but removed it since we don't know how we want to handle batching yet. I can remove anything related to batching from this PR if that makes more sense, since right now we default to one message per batch no matter what

File: src/RabbitMQAttribute.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (katiecai) [reply src/RabbitMQAttribute.cs]: Is there an AutoResolve for type int?

File: src/RabbitMQAttribute.cs (hunk 1 above)
Comment (katiecai) [reply src/RabbitMQAttribute.cs]: Removing exchange for now and setting default value to the empty string

File: src/Trigger/RabbitMQTriggerAttribute.cs [hunk 17]
```diff
@@ -9,24 +9,39 @@ namespace Microsoft.Azure.WebJobs
     [Binding]
     public sealed class RabbitMQTriggerAttribute : Attribute
     {
-        public RabbitMQTriggerAttribute(string hostName, string queueName)
+        public RabbitMQTriggerAttribute(string connectionString, string queueName)
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttribute.cs]: reaname connectionString to connectionStringSetting

File: src/Trigger/RabbitMQTriggerAttribute.cs [hunk 18]
```diff
@@ -9,24 +9,39 @@ namespace Microsoft.Azure.WebJobs
...
-            this.HostName = hostName;
-            this.QueueName = queueName;
-            this.BatchNumber = batchNumber;
+            QueueName = queueName;
         }
 
+        public RabbitMQTriggerAttribute(string hostName, string userName, string password, int port, string queueName)
+        {
+            HostName = hostName;
+            UserName = userName;
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttribute.cs]: Same for username and password: raname to UserNameSetting, PasswordSetting

File: src/Utility.cs [hunk 19]
```diff
@@ -11,5 +12,28 @@ internal static string FirstOrDefault(params string[] values)
...
+                    return true;
+                }
+
+                return false;
+            });
+        }
+
+        internal static bool ValidateUserNamePassword(string userName, string password, string hostName)
+        {
+            if (!hostName.Equals("localhost", StringComparison.InvariantCultureIgnoreCase) && (string.IsNullOrEmpty(userName) || string.IsNullOrEmpty(password)))
```
Comment (pragnagopa) [inline src/Utility.cs]: change references for "localhost" to Constants.LocalHost

File: test/WebJobs.Extensions.RabbitMQ.Tests/UtilityTests.cs [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+using System.Text;
+using Microsoft.Azure.WebJobs.Extensions;
+using Xunit;
+
+namespace WebJobs.Extensions.RabbitMQ.Tests
+{
+    public class UtilityTests
+    {
+        [Fact]
+        public void InvalidCredentials_ReturnsFalse()
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/UtilityTests.cs]: Change this to use theory. Pass expected result as a praram

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+            {
+                Assert.Null(factory.Uri);
+                Assert.Equal("localhost", factory.HostName);
+                Assert.Equal("guest", factory.UserName);
+                Assert.Equal("guest", factory.Password);
+                Assert.Equal(5672, factory.Port);
+            }
+            else if (connectionString != null)
+            {
+                Assert.Equal(new Uri("amqp://user:PASSWORD@52.175.195.81:5672"), factory.Uri);
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: use dummy ip

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+                Assert.Equal("52.175.195.81", factory.HostName);
+                Assert.Equal("user", factory.UserName);
+                Assert.Equal("PASSWORD", factory.Password);
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: rename PASSWORD to testPassoword and rename user to testUserName

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace WebJobs.Extensions.RabbitMQ.Tests
+{
+    public class RabbitMQServiceTests
+    {
+        [Theory]
+        [InlineData(null, "localhost", "queue", "guest", "guest", 5672)]
+        [InlineData("amqp://user:PASSWORD@52.175.195.81:5672", null, "queue", null, null, null)]
+        [InlineData(null, "localhost", "queue", null, null, 0)] // Should fill in "guest", "guest", 5672
+        public void Handles_Connection_Attributes_And_Options(string connectionString, string hostName, string queueName, string userName, string password, int port)
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: pass expected values as part of params. Yes signature will look ugly, but test code will lot simpler as you just need to check for expected values

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs (hunk 21 above)
Comment (katiecai) [reply test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: If I do this the connection will fail - what would be the best solution for this

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs (hunk 22 above)
Comment (pragnagopa) [reply test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: Just test CreateConnectionFactory. do not create RabbitMQ Service

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs (hunk 21 above)
Comment (pragnagopa) [reply test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: Just test CreateConnectionFactory. do not create RabbitMQ Service

Comment (katiecai) [reply test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQServiceTests.cs]: Do I mock RabbitMQService? How do I test this without actually creating the connection since that's what the method does?
